ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/tasks/CompleteSyncBlocksTask.java

```diff
@@ -0,0 +1,180 @@
+import org.hyperledger.besu.ethereum.eth.manager.task.AbstractPeerTask.PeerTaskResult;
+import org.hyperledger.besu.ethereum.eth.manager.task.AbstractRetryingPeerTask;
+import org.hyperledger.besu.ethereum.eth.manager.task.GetSyncBlocksFromPeerTask;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.metrics.BesuMetricCategory;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+import org.hyperledger.besu.plugin.services.metrics.Counter;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+
+import org.apache.tuweni.bytes.Bytes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Given a set of headers, "completes" them by repeatedly requesting additional data (bodies) needed
+ * to create the blocks that correspond to the supplied headers.
+ */
+public class CompleteSyncBlocksTask extends AbstractRetryingPeerTask<List<SyncBlock>> {
+  private static final Logger LOG = LoggerFactory.getLogger(CompleteSyncBlocksTask.class);
+
+  private static final int MIN_SIZE_INCOMPLETE_LIST = 1;
+  private static final int DEFAULT_RETRIES = 5;
+
+  private final EthContext ethContext;
+
+  private final List<BlockHeader> headers;
+  private final Map<Long, SyncBlock> blocks;
+  private final MetricsSystem metricsSystem;
+  private final Counter totalSyncBodiesDownloaded;
+  private final ProtocolSchedule protocolSchedule;
+
+  private CompleteSyncBlocksTask(
+      final ProtocolSchedule protocolSchedule,
+      final EthContext ethContext,
+      final List<BlockHeader> headers,
+      final int maxRetries,
+      final MetricsSystem metricsSystem) {
+    super(ethContext, maxRetries, Collection::isEmpty, metricsSystem);
+    checkArgument(!headers.isEmpty(), "Must supply a non-empty headers list");
+    this.ethContext = ethContext;
+    this.metricsSystem = metricsSystem;
+    this.protocolSchedule = protocolSchedule;
+
+    this.headers = headers;
+    this.blocks =
+        headers.stream()
+            .filter(BlockHeader::hasEmptyBlock)
+            .collect(
+                toMap(
+                    BlockHeader::getNumber,
+                    header ->
+                        new SyncBlock(
+                            header,
+                            createEmptyBodyBasedOnProtocolSchedule(protocolSchedule, header))));
+
+    totalSyncBodiesDownloaded =
+        metricsSystem.createCounter(
+            BesuMetricCategory.SYNCHRONIZER,
+            "total_sync_bodies_downloaded",
+            "Total number of sync bodies downloaded during sync");
+  }
+
+  @Nonnull
+  private SyncBlockBody createEmptyBodyBasedOnProtocolSchedule(
+      final ProtocolSchedule protocolSchedule, final BlockHeader header) {
+    if (isWithdrawalsEnabled(protocolSchedule, header)) {
+      return new SyncBlockBody(
```

Comment:
be nicer to have a static method on SyncBlockBody to create an empty block with withdrawals to hide the RLP encoding being passed in for this case

Reply:
Done